### PR TITLE
 Theoretical Introduction to Linear Regression

### LINKED ISSUE #1164: Theoretical Introduction to Linear Regression
### Brief Summary: 
Describe the what, why, and how of your content idea in 2-5 sentences.
In this article, I discuss the theory behind linear regression. Linear regression is one of the most important algorithms used in machine learning to find the line of best fit given a set of data points. 
### Key Takeaways: 
What are the 3-5 most important things the reader should understand or be able to do after reading this article?
1. What is linear regression?
2. Methods to derive the line of best-fit
- a. Least-Squares solution: One-shot solution
- b. Using Gradients: Iterative solution
3. Code to implement the least-square solution in Python. 
### References: 
N/A
Please list links to any published content/research that you intend to use to support/guide this article. (If none, please indicate 'N/A'.)

### PR DESCRIPTION
**Pull request by lalith1403**
#### Preliminary Checks - Formatting and Structure 
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.

**NOTE: (1st time contributors - submit ONLY one article to get familiar with the process and style-once published, you can then submit any number of articles.)**
      
#### Grammar Checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical Checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution Guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**

This fixes #1164

### CODE REVIEW

**Comment by hectorkambow** [comment]
👋 Thank you for submitting this PR - we will review it ASAP - all PRs go into our queue will be looked over before being approved to be published.

**Review by collins-a — COMMENTED** [review]
Hey @lalith1403, a wonderful article introducing readers to linear regression. Just a few changes and you're good to go.

File: articles/linear-regression-introduction/index.md [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+$$ X^TXb = X^Ty$$
+
+Therefore, the parameter b is found by taking the inverse of $X^TX$ on both sides. Therefore we obtain, 
+
+$$ b= (X^TX)^{-1}X^TY $$
+
+This can be solved using many methods, and the solution is computed. Some of the methods that can be used to solve the same are:
+- QR decomposition
+- Moore - Penrose inverse
+- SVD decomposition
```
[the remark this answers is not in the document]
Comment (lalith1403) [reply articles/linear-regression-introduction/index.md:68]: > I think a brief description of these methods would be valuable to readers.

I don't think I can explain these methods in one or two lines. They are detailed concepts. I will simply mention that these are methods to compute inverses in various situations,

File: articles/linear-regression-introduction/index.md [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    def fit(self,X,y):
+        self.weights = np.dot(np.linalg.inv(np.dot(X.T, X)), np.dot( X.T, y ))
+        self.bias = y.mean() - np.sum(self.weights * X.mean(axis=0))
+    
+    def predict(self,X):
+        return np.dot(X, self.weights )+ self.bias
+    
+    def score(self,X,y):
+        return 1-(np.sum(((y-self.predict(X))**2))/np.sum((y-np.mean(y))**2))
+```
```
[the remark this answers is not in the document]
Comment (lalith1403) [reply articles/linear-regression-introduction/index.md:91]: Resolved this by adding a seed value during train_test_split. This way, results are reproducible.

**Comment by lalith1403** [comment]
@collins-a I have made the required changes.

**Comment by hectorkambow** [comment]
👍  
Great article @lalith1403  
I would probably advise to make the intro a bit more introductory - just a suggestion: "This article will cover - or in this article we will discuss" 

May lend itself to flow a bit easier for the audience. 

Please see the minor if any edits I made cc: @collins-a  

Added a section link and the mathjax script for the $$ equations 🚀
